### DISCUSSION TITLE
Proposal: Migrate web site to markdown and mkdocs

### DISCUSSION THREAD

**Discussion by reece** [HGVSnomenclature/hgvs-nomenclature#17] [HGVS presentation]
The current [VarNomen](https://github.com/HGVSnomenclature/VarNomen) web site uses Jekyll with html files, extensive use of embedded HTML within markdown, and extensive custom formatting and layout. These factors thwart editing and contributing. This discussion is a proposal to migrate the existing website to mkdocs, a modern static site generator, and to convert html files and embedded html to markdown where practical. It is expected that this migration will facilitate updates and contributions, thereby relieving a significant burden for Johan.

# Scope

## In-scope

- Migrate all pages as-is to mkdocs. Limit content changes except as required to achieve workable pages.
- Mechanism to deploy pages on github.
- Versioning mechanism that enables historical versions to be viewed

## Out-of-Scope

- Restructuring pages will be taken up separately after migration. (Some restructuring is inevitable in order to migrate from Jekyll to mkdocs.)

# Implementation

Implementation is in the [17-migrate-web-site-to-markdown-and-mkdocs](https://github.com/HGVSnomenclature/HVNC/tree/17-migrate-web-site-to-markdown-and-mkdocs) branch. Up-to-date builds may be seen at https://hgvs-nomenclature.org/en/17-migrate-web-site-to-markdown-and-mkdocs/ .

## Phase 0: Proof-of-concept  

The first phase of this work will be to implement a mkdocs framework for pages and migrate 1-5 pages for review and decision by the HVNC group. Investigate versioning options (likely either [mike](https://github.com/jimporter/mike) or [read the docs](https://readthedocs.org/).  

## Phase 1: Consolidate above repos into single repo

- [x] Johan grants Reece owner access to the organization. (see below)
- [x] Reece archives the VarNomen pages so that they are available for historical purposes, but not writeable. NO UPDATES TO VARNOMEN AFTER THIS TIME.
- [x] ~~Reece imports the (now archived) VarNomen web pages~~ Decided against this. Let the VarNomen pages stand as-is.
- [x] Reece migrates the [17-migrate-web-site-to-markdown-and-mkdocs](https://github.com/hgvs-nomenclature/hgvs-nomenclature.github.io/tree/17-migrate-web-site-to-markdown-and-mkdocs) branch with the new pages
- [x] Reece removes the temporary hgvs-nomenclature/hgvs-nomenclature repo (which is no longer needed).
- [x] Reece renames the HGVC repo to hgvs-nomenclature.

## Phase 2: Automated migration of existing varnomen pages

The goal of this phase is to undertake a first pass at migrating all of the varnomen pages to mkdocs, make links work, establish automated builds, etc. **All stylistic and manual changes are out of scope for this phase.**

- [x] Reece arranges for read the docs updates from this branch.
- [x] Reece finishes the migrating VarNomen pages and submits a PR.
- [x] We merge the PR. The goal for this PR is only to achieve parity with the original website content.

At this point, the new pages will be at hgvs-nomenclature.org (but essentially unknown to the world) and the old pages will still appear at varnomen.hgvs.org.

## Phase 3: Updating the website

- [x]  Manual updates to website. See [HVNC Board](https://github.com/orgs/HGVSnomenclature/projects/3/views/1) for specific tasks.

## Phase 4: Cutover to new site

- [ ]  Create a forwarding page at varnomen.hgvs.org announcing the new site. Visitors to varnomen.hgvs.org will be automatically redirected by the web page (after a short announcement).
- [ ] Someone (Johan? Reece?) asks hgvs.org admins to permanently redirect varnomen.hgvs.org.

**Comment by reece** [comment]
# Re Q1

Reece's recommendation: "HGVS Nomenclature" (slug: `hgvs-nomenclature`). The organization should use that name (i.e., "HGVSnomeclature" → "hgvs-nomeclature"), the committee repo should remain as HVNC, and we'll create a new repo, "hgvs-nomenclature.github.io" to deploy github pages.

**Comment by reece** [comment]
In the 17-migrate-web-site-to-markdown-and-mkdocs branch, I:

- setup an initial mkdocs structure
- migrated the background and recommendations as-is (_bg-material and _recommendations in VarNomen repo) 

The result looks like this:

![image](https://user-images.githubusercontent.com/109453/214442179-0552d3c9-b1ce-4ef5-8307-f6045032246b.png)

@ahwagner

**Comment by reece** [comment]
Update:
- created hgvs-nomenclature github org
- forked HVNC to hgvs-nomenclature/hgvs-nomenclature.github.io
- setup read the docs to build docs automatically (on existing branches). The result is at https://hgvs-nomenclature.readthedocs.io/en/17-migrate-web-site-to-markdown-and-mkdocs/.
- bought https://hgvs-nomenclature.org and pointed it at the above 

@ahwagner @jfjlaros @jtdendunnen: I like this direction but want feedback before I get too far ahead.

If we end up keeping this work, I'll arrange to share the admin rights.

**Reply by ifokkema** [reply]
Out of curiosity, why not simply [rename the current organization](https://docs.github.com/en/organizations/managing-organization-settings/renaming-an-organization) instead of starting over completely? Or is this just temporary?

Otherwise, it looks great so far!

**Reply by ahwagner** [reply]
I think @reece needed broader permissions to demonstrate the proposed system, and since this is for demo purposes decided to just do it in a new workspace where he could put all the pieces together without waiting on e.g. admin permissions.

I think this is a big step forward and I'm glad that we're using the versioning tools that RTD provides. Thanks for putting this together @reece!

Another feature to showcase might be the RTD auto-build on PRs as well.

**Reply by reece** [reply]
Exactly. I didn't have permissions and was eager to proceed. To be clear: I don't need to own this and will be happy to spread the admin load around to people who know how, and to rename and/or transfer repos as the group deems appropriate.

**Comment by reece** [comment]
P.S. Notice the version menu in the lower right corner. Right now it shows only latest (which is a template) and the 17- branch. Eventually, it will provide navigation to other versions of the documentation. Through this mechanism we will be able to provide links to historical versions and to the github repo.
![image](https://user-images.githubusercontent.com/109453/214731242-93853c6b-53c3-4245-ba60-451074adf78a.png)

**Comment by reece** [comment]
All VarNomen pages have been migrated. The structure and content are identical to varnomen. 

The mobile site is also worth checking out. See here, with the nav menu expanded:
![image](https://user-images.githubusercontent.com/109453/215308209-54c2d312-e2b9-4ba9-8767-e5e6f897052d.png)

@ahwagner @ifokkema @jfjlaros @Peter-J-Freeman @murphyte  @jtdendunnen

**Reply by ifokkema** [reply]
Great work! What would be the next steps? I see some pages will need to be re-formatted, and links should be corrected, but we can get that done later.

**Reply by marinadistefano** [reply]
Agree with everyone else, this looks great (save for the issues with the variant sections)! While I don't have any website expertise, I'm happy to spend some time looking through the pages for typos etc when it's all migrated correctly. I think I found 1 or 2 just randomly clicking through the site.

**Comment by jtdendunnen** [comment]
nice work, what do you suggest should be next?
As far as I can see some things now appear in a strange format; The old site has for DNA > Substitution several main sections (Definition, Examples, Q&A) which on the new site are split over different pages. Should these not be merged into one page?

**Comment by reece** [comment]
@ifokkema @marinadistefano @jtdendunnen Thanks for above comments. I have made a significant push to fix some formatting issues, ensure all content is migrated, and that nearly all links work. Please review for glaring issues.

**Comment by reece** [comment]
The website overhaul has made significant progress and is now ready for feedback. The current prerelease version of the HGVS Nomenclature is tagged 24.1.0 (intending to release in January). Please review https://hgvs-nomenclature.org/ and submit issues to https://github.com/HGVSnomenclature/hgvs-nomenclature/issues (preferred, or send email to reecehart@gmail.com if necessary).


## Content changes
- With the exceptions in the next bullets, I avoided significant content changes. The sentences and sections in the new pages are nearly verbatim from the VarNomen pages. We can undertake rearrangements and rewording later if desired.
- Minor changes to the site topology. E.g., history and versioning are now in the background section.
- Moved contact us to Home page (https://hgvs-nomenclature.org/). Marked the Facebook group as obsolete. Added GitHub discussions and new email group.
- Described proposal process in https://hgvs-nomenclature.org/consultation/
- Described style conventions in https://hgvs-nomenclature.org/recommendations/style/
- Adopted semantic versioning conventions in https://hgvs-nomenclature.org/background/versioning/
- Added explainer for EBNF-like syntax in https://hgvs-nomenclature.org/recommendations/grammar/
- Added a new HGVS Summary table in https://hgvs-nomenclature.org/recommendations/summary/
- Merged recent corrections from Johan (various pages)
- Added new [software](https://hgvs-nomenclature.org/software/) page with solicitation for additions

## Visual changes
- The syntax now uses a style that is conventional for computer scientists, but presented in a way that, I hope, remains accessible to readers. See https://hgvs-nomenclature.org/recommendations/grammar/.
- HGVS expressions (and fragments of them) use a consistent typeface. There is now a way to render counterexamples (in a red type face) to denote invalid expressions. See https://hgvs-nomenclature.org/recommendations/style/.
- The pages generally have a cleaner layout.
- The outline menu provides a good overview of the pages in one glance.
- Readers will be able to browse historical releases and prereleases as they become available. (Historical releases of the varnomen pages are not included.)
- Ability to select HGVS Nomenclature version in browser, like this:
![image](https://github.com/HGVSnomenclature/hgvs-nomenclature/assets/109453/602ba1fd-727b-49c5-93d2-798d515994a9)

## Maintainer features (invisible to the user)
- The most significant and beneficial change is that all syntax expressions are consolidated in a single file, alongside examples and explanatory text. Unlike the previous pages, these blocks are rendered programmatically (rather than manually), which ensures consistent formatting and enables alternate views without adding maintenance. For example, the syntax file generates a helpful table of all HGVS expressions on one page. See https://hgvs-nomenclature.org/recommendations/summary/. 
- The page content is nearly exclusively markdown. This improves consistency and is easier to maintain. HTML is used sparingly, notably for complex tables and for styled HGVS expressions.
- The syntax table includes a glossary of common symbols. Symbols relevant to a specific page are included automatically.

## Remaining Required and Desired Work

See https://github.com/orgs/HGVSnomenclature/projects/3


@ahwagner @ifokkema @jtdendunnen @jtdendunnen @Peter-J-Freeman @marinadistefano

**Reply by rileygr** [reply]
@reece I've reviewed this and it's a terrific improvement.  Thanks for your work on this.  I particularly like the little versioning widget, however it might be nice to add a little explanation of it and how it works.  It's not very intuitive, e.g., whether the links there are for the current page, what the various links go to, what patch "0rc1" in "24.1.0rc1" is, etc.  

![image](https://github.com/HGVSnomenclature/hgvs-nomenclature/assets/5871381/9b44184e-e415-4012-a151-e6065567c2a9)

**Reply by reece** [reply]
Thanks for the feedback @rileygr. That widget is from readthedocs.io, so we don't have much control over it.

**Reply by reece** [reply]
(And thanks for the positive feedback!)

**Comment by rileygr** [comment]
@reece, Today is apparently my day for commenting.  In reviewing some ClinVar validation, I saw that we accept alleles like "sequence_identifier ":" coordinate_type ".**variant1(;)variant2”, i.e. without using “[ ]”**, as seen on the [current varnomen allele page](http://varnomen.hgvs.org/recommendations/DNA/variant/alleles/) rather than "sequence_identifier ":" coordinate_type ".**[" position_edit "(;)" position_edit "]"** " with "[ ]" as seen on the [new hgvs-nomenclature allele page](https://hgvs-nomenclature.org/en/latest/recommendations/DNA/alleles/).  Is this an intentional change that we should code against?

**Reply by reece** [reply]
@rileygr : Would you please double check? I think the syntax and examples are nearly identical in the old and new pages. For better or worse, both versions show syntax only **with** brackets, but there are examples **without** brackets in the bottom set of examples.

Do I misunderstand the point you raise? If this needs more discussion, would you please create a separate discussion thread? I'd like to keep this thread just about the migration.

**Reply by ifokkema** [reply]
He's right. The new page indeed states `NC_000001.11:g.[123G>A(;)345del]` as a syntax example. The original page never had this example and only listed _cis_ (`[...;...]`) and _trans_ (`[...];[...]`) as possible syntax, with an example in the notes of "variant1(;)variant2". This latter example and many other notes have been deleted on the new page and are no longer present. A large number of pull requests have been merged without proper review in the last few weeks, but I don't know if this change has been done in this period. If you go through the history, you'll probably quickly find the issue.
In short, the new pages are wrong; they contain new syntax that hasn't been agreed on and important stuff got removed.

**Reply by reece** [reply]
Thanks for the report, @rileygr, and the clarification @ifokkema .

I created https://github.com/HGVSnomenclature/hgvs-nomenclature/issues/108 and addressed it in https://github.com/HGVSnomenclature/hgvs-nomenclature/pull/109.

**Reply by ifokkema** [reply]
@reece Thanks for the fix! ~~I guess we'll now either need to change the `[...(;)...]` to `...(;)...` in the syntax block, or we need to discuss this in the next meeting. I personally agree that having square brackets around the variants makes the most sense, so I suggest trying to push it into the next meeting.~~

**Reply by ifokkema** [reply]
@reece Ah sorry, I see now the PR is still open, and you have fixed the syntax blocks as well. I had checked the website instead. :joy:

**Reply by jtdendunnen** [reply]
Originally, the HGVS recommended format was set to "variant1(;)variant2”, i.e. without using “[ ]” since the addition of "[]" adds no information. From "(;)" is is already clear that it is not known whether the variants are on the same allele or not. A general rule is that we do not include redundant information in variant descriptions (so not g.123delA but g.123del). Furthermore, when more variants are involved, the description with "[]' becomes very complex.

**Comment by rileygr** [comment]
@reece I'm not suggesting you change the widget, just maybe add a couple of words explaining its use.
